### DISCUSSION TITLE
Error in Virtual User 2: can't read "RAISEERROR": no such variable

### DISCUSSION THREAD

**Discussion by dylan2intel** [TPC-Council/HammerDB#364] [Q&A] [answered]
```
After enabled below variables
diset tpcc mysql_async_scale ${TPCC_ASYNC_SCALE:-false}
diset tpcc mysql_connect_pool ${TPCC_CONNECT_POOL:-false}

...
Vuser 125 created - WAIT IDLE
17: 
Vuser 126 created - WAIT IDLE
17: 
Vuser 127 created - WAIT IDLE
17: 
Vuser 128 created - WAIT IDLE
17: 
Vuser 129 created - WAIT IDLE
17: 
Logging activated
17: to /tmp/hammerdb_6292201A5E0103E233737303.log
17: 
129 Virtual Users Created with Monitor VU
17: 
Vuser 1:RUNNING
17: 
Vuser 1:Beginning rampup time of 2 minutes
17: 
Vuser 2:RUNNING
17: 
Vuser 2:Started asynchronous clients:vuser2:ac1 vuser2:ac2 vuser2:ac3 vuser2:ac4 vuser2:ac5 vuser2:ac6 vuser2:ac7 vuser2:ac8 vuser2:ac9 vuser2:ac10
17: 
Vuser 3:RUNNING
17: 
Vuser 3:Started asynchronous clients:vuser3:ac1 vuser3:ac2 vuser3:ac3 vuser3:ac4 vuser3:ac5 vuser3:ac6 vuser3:ac7 vuser3:ac8 vuser3:ac9 vuser3:ac10
17: 
Vuser 4:RUNNING
17: 
Vuser 4:Started asynchronous clients:vuser4:ac1 vuser4:ac2 vuser4:ac3 vuser4:ac4 vuser4:ac5 vuser4:ac6 vuser4:ac7 vuser4:ac8 vuser4:ac9 vuser4:ac10
17: 
Error in Virtual User 2: can't read "RAISEERROR": no such variable
17: 
Vuser 2:FINISHED FAILED
17: 
Vuser 5:RUNNING
...
```

**Comment by dylan2intel** [comment]
Refer to https://github.com/TPC-Council/HammerDB/blob/master/config/mysql.xml

```
<mysql_async_scale>false</mysql_async_scale>
<mysql_async_client>10</mysql_async_client>
<mysql_async_verbose>false</mysql_async_verbose>
<mysql_async_delay>1000</mysql_async_delay>
<mysql_connect_pool>false</mysql_connect_pool>
```

The default value of mysql_async_client is 10, is above issue regarding to the client count?

**Comment by sm-shaw** [comment]
This is not correct 

```
diset tpcc mysql_async_scale ${TPCC_ASYNC_SCALE:-false}
diset tpcc mysql_connect_pool ${TPCC_CONNECT_POOL:-false}

```
settings should be true or false

**Reply by dylan2intel** [reply]
> This is not correct
> 
> ```
> diset tpcc mysql_async_scale ${TPCC_ASYNC_SCALE:-false}
> diset tpcc mysql_connect_pool ${TPCC_CONNECT_POOL:-false}
> ```
> 
> settings should be true or false

Actually, it is true or false, I just wrap it as a variable in shell, which will auto-gen a .tcl file

**Comment by sm-shaw** [comment]
Try troubleshooting first without your wrapper script. 
What does the mysqlcpool.xml look like? To use the connect pool this file must be edited for all of the connections you want to configure and which connections run which stored procedures.

**Comment by dylan2intel** [comment]
Could you pls. show an example of mysqlcpool..xml, since my hammerdb program entrance is hammerdbcli as follows, how can I edit the mysqlcpool.xml?

echo "===Stage 1: Build schema started==="
./hammerdbcli auto ${TPCC_TCL_SCRIPT_PATH}/build_schema.tcl | tee /build_schema_${DB_TYPE}_tcl.log
echo "===Stage 1: Build schema finished==="

echo "===Stage 2: Run timer started"
$NUMACTL_OPTIONS ./hammerdbcli auto ${TPCC_TCL_SCRIPT_PATH}/run_timer.tcl | tee /run_timer_${DB_TYPE}_tcl.log
echo "===Stage 2: Run timer finished"

**Comment by sm-shaw** [comment, marked as answer]
OK, there is an issue here that needs to be fixed,  that is triggered by an error in configuration.  So if you fix the configuration error, then it will not cause the issue. 
The error message you see is that an error is triggered, however it cannot see the RAISEERROR variable so you see this error first before the actual error - this is what will be fixed. 
What is most likely the cause is that you trigger the Too many connections MySQL error as follows: 
`Error in Virtual User 6: vuser6:ac10:socket login failed:mysqlconnect/db server: Too many connections`
This is because with the connect pool and asynchronous clients each asynchronous client will create all of the connect pool connections acting as a multiplier therefore you will create a large number of sessions even with a relaltively small number of virtual users. 

To edit the connect pool parameters simply use any text editor. You would want to use this typically in a clustered environment, however for testing as an example below I have already created 3 databases on the same server and my transactions will round-robin between them. 
```
<connpool>
<connections>
    <c1>
	<mysql_host>127.0.0.1</mysql_host>
        <mysql_port>3306</mysql_port>
        <mysql_socket>/tmp/mysql.sock</mysql_socket>
	<mysql_user>root</mysql_user>
        <mysql_pass>mysql</mysql_pass>
        <mysql_dbase>tpcc1</mysql_dbase>
    </c1>
    <c2>
	<mysql_host>127.0.0.1</mysql_host>
        <mysql_port>3306</mysql_port>
        <mysql_socket>/tmp/mysql.sock</mysql_socket>
	<mysql_user>root</mysql_user>
        <mysql_pass>mysql</mysql_pass>
        <mysql_dbase>tpcc2</mysql_dbase>
    </c2>
    <c3>
	<mysql_host>127.0.0.1</mysql_host>
        <mysql_port>3306</mysql_port>
        <mysql_socket>/tmp/mysql.sock</mysql_socket>
	<mysql_user>root</mysql_user>
        <mysql_pass>mysql</mysql_pass>
        <mysql_dbase>tpcc3</mysql_dbase>
    </c3>
</connections>
<sprocs>
  	<neworder>
		<connections>c1 c2 c3</connections>
    		<policy>round_robin</policy>
	</neworder>
    	<payment>
		<connections>c1 c2 c3</connections>
    		<policy>round_robin</policy>
	</payment>
    	<delivery>
		<connections>c1 c2 c3</connections>
    		<policy>round_robin</policy>
	</delivery>
    	<stocklevel>
		<connections>c1 c2 c3</connections>
    		<policy>round_robin</policy>
	</stocklevel>
    	<orderstatus>
		<connections>c1 c2 c3</connections>
    		<policy>round_robin</policy>
	</orderstatus>
</sprocs>
</connpool>
```

So if you configure the connect pool correctly and increase your max_connections above the number of sessions you want to allow to connect (and prepared statements is likely to be another limit). Then you should not see the issue error. 
```
mysql> show global variables like 'max_connect%';
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| max_connect_errors | 100   |
| max_connections    | 200   |
+--------------------+-------+
2 rows in set (0.00 sec)


mysql> show status where variable_name = 'threads_connected';
+-------------------+-------+
| Variable_name     | Value |
+-------------------+-------+
| Threads_connected | 201   |
+-------------------+-------+
1 row in set (0.00 sec)


mysql> show global variables like 'max_prepared%';
+-------------------------+-------+
| Variable_name           | Value |
+-------------------------+-------+
| max_prepared_stmt_count | 500   |
+-------------------------+-------+
1 row in set (0.01 sec)

mysql> set global max_prepared_stmt_count = 16384;
Query OK, 0 rows affected (0.00 sec)

mysql> show global variables like 'max_prepared%';
+-------------------------+-------+
| Variable_name           | Value |
+-------------------------+-------+
| max_prepared_stmt_count | 16384 |
+-------------------------+-------+
1 row in set (0.01 sec)
```

However, the issue in HammerDB is shown in the example below. In ConnectToMySQLAsynch if it catches an error it checks the RAISEERROR variable that is not set local to the proc - so this is why the "can't read "RAISEERROR": no such variable" error is seen first.  As we have multiple asynchronous clients and don't want it to error here just report the error then the if $RAISEERROR can be removed . 

```
proc ConnectToMySQLAsynch { host port socket user password db clientname async_verbose } {
    global mysqlstatus
    if { [ chk_socket $host $socket ] eq "TRUE" } {
        if [catch {mysqlconnect -socket $socket -user $user -password $password} mysql_handler] {
            set connected "FALSE"
            if { $RAISEERROR } {
               puts "$clientname:socket login failed:$mysqlstatus(message)"
		}
        } else {
            set connected "TRUE"
        }
...
```

**Reply by sm-shaw** [reply]
Create Issue #365
